src/natcap/invest/validation.py

```diff
@@ -1005,11 +987,30 @@ def validate(args, spec, spatial_overlap_opts=None):
         # Extra args that don't exist in the MODEL_SPEC are okay
         # we don't need to try to validate them
         try:
-            parameter_spec = spec[key]
+            # Using deepcopy to make sure we don't modify the original spec
+            parameter_spec = copy.deepcopy(spec[key])
         except KeyError:
             LOGGER.debug(f'Provided key {key} does not exist in MODEL_SPEC')
             continue
 
+        # rewrite parameter_spec for any nested, conditional validity
+        axis_keys = None
+        if parameter_spec['type'] == 'csv':
+            axis_keys = ['columns', 'rows']
+        elif parameter_spec['type'] == 'vector':
+            axis_keys = ['fields']
+
```

Comment:
Could we also include `'directory'` here with its key `'contents'`? This isn't used much, but would be nice to have a comprehensive solution

Reply:
Happy to!  Although it looking more closely at `check_directory`, we don't currently do anything with the `contents` key.  The main `validation` loop will probably need a bit of a refactor to support this kind of recursion.  880e31fbd